R/spatial_clustering_cv.R

```diff
@@ -32,17 +38,42 @@
 #' pp. 5372-5375, doi: 10.1109/IGARSS.2012.6352393.
 #'
 #' @examples
-#' data(ames, package = "modeldata")
-#' spatial_clustering_cv(ames, coords = c(Latitude, Longitude), v = 5)
+#' data(Smithsonian, package = "modeldata")
+#' spatial_clustering_cv(Smithsonian, coords = c(latitude, longitude), v = 5)
+#'
+#' smithsonian_sf <- sf::st_as_sf(Smithsonian,
+#'                                coords = c("longitude", "latitude"),
+#'                                # Set CRS to WGS84
+#'                                crs = 4326)
+#'
+#' # When providing sf objects, coords are inferred automatically
+#' spatial_clustering_cv(smithsonian_sf, v = 5)
+#'
+#' @rdname spatial_clustering_cv
 #' @export
-spatial_clustering_cv <- function(data, coords, v = 10, ...) {
-  coords <- tidyselect::eval_select(rlang::enquo(coords), data = data)
+spatial_clustering_cv <- function(data, coords, v = 10, cluster_function = c("kmeans", "hclust"), ...) {
+  cluster_function <- rlang::arg_match(cluster_function)
 
-  if (is_empty(coords)) {
-    rlang::abort("`coords` are required and must be variables in `data`.")
+  if ("sf" %in% class(data)) {
+    if (!missing(coords)) {
+      rlang::warn("`coords` is ignored when providing `sf` objects to `data`.")
+    }
+    coords <- sf::st_centroid(sf::st_geometry(data))
+    coords <- as.dist(sf::st_distance(coords))
+  } else {
+    coords <- tidyselect::eval_select(rlang::enquo(coords), data = data)
+    if (is_empty(coords)) {
+      rlang::abort("`coords` are required and must be variables in `data`.")
+    }
+    coords <- data[coords]
+    coords <- dist(coords)
```

Comment:
This chunk isn't too long, but we could make a new generic function here to compute the `dist` (possible name such as `compute_coords_dist()`), with methods for a dataframe and `sf`. Do you think we would use something like this in another method as well? Or should we just stick with `if () {} else {}`?

Reply:
I don't _think_ we'll need this exact method again. I _believe_ the other methods we've talked about are going to rely pretty heavily on `sf` distance calculations, and less on the distance matrix for the entire data set, but I could be wrong there. Happy to extract it if we want though!

Reply:
Let's leave it as is for now. 👍